rfd/0074-sftp-support.md

```diff
@@ -0,0 +1,136 @@
+---
+authors: Andrew LeFevre (andrew.lefevre@goteleport.com)
+state: draft
+---
+
+# RFD 74 - SFTP Support
+
+## Required Approvers
+
+- Engineering: `@r0mant && @jakule`
+- Product: `@klizhentas || @xinding33`
+
+## What
+
+Add SFTP support to `tsh` and Teleport Node and Proxy services. This will
+not replace scp, both will be supported.
+
+## Why
+
+[OpenSSH 9.0](https://www.openssh.com/txt/release-9.0) changed `scp` to
+use the [SFTP](atatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-02)
+protocol by default instead of scp/rcp protocol. This is to
+combat many security issues present in the legacy scp/rcp protocol. Users
+running the latest OpenSSH version will currently find that `scp` and various
+tools that use it under the hood, such as Ansible and various JetBrain IDEs
+will not work through Teleport anymore. This can be fixed temporarily by passing
+a flag to `scp`, but that can be difficult when `scp` is being invoked indirectly
+by other tools.
+
+## Details
+
+[github.com/pkg/sftp](https://pkg.go.dev/github.com/pkg/sftp) will be leveraged 
+heavily to provide both an SFTP client and server.
+
+SFTP support will be added in three stages. 
+
+1. Adding SFTP subsystem support to Teleport's SSH server.
+When an SFTP request is first received on an SSH connection, the Teleport daemon
+will be re-executed as the logon user of the SSH connection. This will require
+adding a hidden `sftp` sub command to the `teleport` binary. The parent process will
+create 2 anonymous pipes and pass them to the child (`teleport sftp`) so the child
+can access the SFTP connection. The `enable_file_copying` option will be added to
+the `ssh_server` yaml config to control whether scp and sftp will be enabled or not
+for per node.
+
+2. Adding SFTP protocol support to `tsh scp` sub command, and
+ensuring SFTP trasfers work on the web UI. `tsh scp` will continue to use the scp/rcp
+protocol by default for backwards compatibility, but the SFTP protocol can be
+optionally enabled with the `-s` flag.
+
+3. Making the SFTP protocol be used by default for
+`tsh scp`, and adding the `-O` flag to allow optional usage of thee scp/rcp protocol
+for backwards compatibility. This stage will take place in a future major release
+(11?) and will be documented as a potentially breaking change.
+
+Both the `-s` and `-O` flags are what OpenSSH uses to allow the user to choose
+what protocol to use, and the deprecation process of making SFTP optional then
+default is what OpenSSH did as well.
+
+A potential fourth stage would be adding the `tsh sftp` sub command which would
+be very similar in behavior to OpenSSH's `sftp` command. This will only be done
+if users express a need for it.
+
+#### Roles
+
+A role option `disable_file_copy` will be added that will define if file
```

Comment:
To be consistent with example below:
```suggestion
A role option `disable_copy` will be added that will define if file
```

Reply:
Oops, copy-pasta oversight. I personally prefer `disable_file_copy` because it's explicit what kind of copying is being controlled.

Reply:
Oops, copy-pasta oversight. I personally prefer disable_file_copy because it's explicit what kind of copying is being controlled.

Reply:
Oops, copy-pasta oversight. I personally prefer disable_file_copy because it's explicit what kind of copying is being controlled.